tests/csapi/sync_archive_test.go

```diff
@@ -0,0 +1,436 @@
+		Content: map[string]interface{}{
+			"my_key": "after",
+		},
+	})
+
+	exhaustiveResCheck := func(result gjson.Result, origin string) {
+		roomRes := result.Get("rooms.leave." + client.GjsonEscape(roomID))
+
+		stateEvents := roomRes.Get("state.events")
+		if stateEvents.Exists() {
+			if !stateEvents.IsArray() {
+				t.Fatalf("state.events is not an array")
+			}
+			if len(stateEvents.Array()) != 0 {
+				t.Fatalf("expected no state events in %s", origin)
+			}
+		}
+
+		timelineEvents := roomRes.Get("timeline.events")
+		if !timelineEvents.Exists() {
+			t.Fatalf("timeline.events does not exist in %s", origin)
+		}
+		if !timelineEvents.IsArray() {
+			t.Fatalf("timeline.events is not an array")
+		}
+		if len(timelineEvents.Array()) != 2 {
+			t.Fatalf("Expected two timeline events in %s", origin)
+		}
+
+		timelineEvent := timelineEvents.Array()[0]
+		if timelineEvent.Get("content.body").Str != "before" {
+			t.Fatalf("Expected only events from before leaving in %s", origin)
+		}
+	}
+
+	syncRes, _ := bob.MustSync(t, client.SyncReq{Filter: bobFilter})
+
+	exhaustiveResCheck(syncRes, "syncRes")
+
+	sinceSyncRes, _ := bob.MustSync(t, client.SyncReq{Filter: bobFilter, Since: bobSince})
+
+	exhaustiveResCheck(sinceSyncRes, "sinceSyncRes")
+}
+
+// This tests if rooms in the leave section get removed after a sync.
+//
+// sytest: Previously left rooms don't appear in the leave section of sync
+func TestOlderLeftRoomsNotInLeaveSection(t *testing.T) {
+	runtime.SkipIf(t, runtime.Dendrite) // FIXME: https://github.com/matrix-org/dendrite/issues/1323
+
+	deployment := Deploy(t, b.BlueprintOneToOneRoom)
+	defer deployment.Destroy(t)
+
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	bob := deployment.Client(t, "hs1", "@bob:hs1")
+
+	aliceFilter := createFilter(t, alice, map[string]interface{}{
+		"room": map[string]interface{}{
+			"timeline": map[string]int64{
+				"limit": 1,
+			},
+			"include_leave": true,
+		},
+	})
+
+	roomToLeave := alice.CreateRoom(t, map[string]string{
+		"preset": "public_chat",
+	})
+	roomToSpam := alice.CreateRoom(t, map[string]string{
+		"preset": "public_chat",
+	})
+
+	bob.JoinRoom(t, roomToLeave, nil)
+	bob.JoinRoom(t, roomToSpam, nil)
+	bob.MustSyncUntil(
+		t,
+		client.SyncReq{},
+		client.SyncJoinedTo(bob.UserID, roomToLeave),
+		client.SyncJoinedTo(bob.UserID, roomToSpam),
+	)
+
+	_, aliceSince := alice.MustSync(t, client.SyncReq{Filter: aliceFilter, TimeoutMillis: "0"})
+
+	alice.LeaveRoom(t, roomToLeave)
+
+	aliceSince = alice.MustSyncUntil(t, client.SyncReq{Filter: aliceFilter, Since: aliceSince}, client.SyncLeftFrom(alice.UserID, roomToLeave))
+
+	bob.SendEventSynced(t, roomToLeave, b.Event{
+		Type:     "m.room.member",
+		StateKey: b.Ptr(bob.UserID),
+		Content: map[string]interface{}{
+			"membership": "join",
+			"filler":     "Welcome back, my friends, to the show that never ends...",
+		},
+	})
+
+	for i := 0; i < 20; i++ {
+		alice.SendEventSynced(t, roomToSpam, b.Event{
```

Comment:
nit: in the sytest, it's bob who sends the spam. It looks like it doesn't really matter though.
```suggestion
		bob.SendEventSynced(t, roomToSpam, b.Event{
```